File: scripts/python/distancia_de_mahalanobis_y_z_score.py
import numpy as np
import pandas as pd
from scipy.spatial.distance import mahalanobis

df = df

columnas_logicas = df.select_dtypes(include="bool").columns
df[columnas_logicas] = df[columnas_logicas].astype(float)
df = df.fillna(0)

datos = df.iloc[:, 1:]
medias = datos.mean()
covarianzas = datos.cov()
inv_cov = np.linalg.inv(covarianzas.values)

df["distancia_mahalanobis"] = datos.apply(lambda fila: mahalanobis(fila, medias, inv_cov), axis=1)

media_dist = df["distancia_mahalanobis"].mean()
sd_dist = df["distancia_mahalanobis"].std()
df["z"] = (df["distancia_mahalanobis"] - media_dist) / sd_dist

df.to_csv("resultado_mahalanobis.csv", sep=";")

File: scripts/r/distancia_de_mahalanobis_y_z_score.R
library(dplyr)
library(stats)

df <- df

columnas_logicas <- names(df)[sapply(df, is.logical)]
df[columnas_logicas] <- lapply(df[columnas_logicas], as.numeric)
df[is.na(df)] <- 0

medias <- colMeans(df[, -1])
covarianzas <- cov(df[, -1])

distancias_mahalanobis <- apply(df[, -1], 1, function(x) mahalanobis(x, medias, covarianzas))

df$distancia_mahalanobis <- distancias_mahalanobis

media_dist <- mean(df$distancia_mahalanobis)
sd_dist <- sd(df$distancia_mahalanobis)
df$z <- (df$distancia_mahalanobis - media_dist) / sd_dist

write.csv2(df, "resultado_mahalanobis.csv")
